File: py/balanced_binary_tree/balanced_binary_tree.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def get_height(self, element: Optional[TreeNode]) -> int:
        if element is None:
            return 0


    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        if abs(self.get_height(root.left) - self.get_height(root.right)) > 1:
            return False
        return True

File: go/balancedbinarytree/balanced_binary_tree.go
package balancedbinarytree

func isBalanced(root *TreeNode) bool {
	_, b := isBalancedHeight(root)
	return b
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func isBalancedHeight(root *TreeNode) (int, bool) {
	if root == nil {
		return 0, true
	}

	lh, bl := isBalancedHeight(root.Left)
	if !bl {
		return 0, false
	}

	rh, br := isBalancedHeight(root.Right)
	if !br {
		return 0, false
	}
	if abs(lh-rh) > 1 {
		return 0, false
	}
	return max(lh, rh) + 1, true
}
